### PR TITLE
feat(cloudwatch): Add L2 for Insight Rule + add contributor insights widget

### PR DESCRIPTION
**Pull request by seanyles-amzn**
Adds an L2 Insight Rule and a corresponding Insight Widget.  My team had a use for it so I wrote one up and thought I'd PR it.

This PR continues PRs [19534](https://github.com/aws/aws-cdk/pull/19534) and [13008](https://github.com/aws/aws-cdk/pull/13008).

### CODE REVIEW

**Review by TheRealAmazonKendra — DISMISSED** [review]
New L2 constructs require an approved [RFC](https://github.com/aws/aws-cdk-rfcs) before a pull request will be reviewed. See our [state diagram for the RFC Process](https://github.com/aws/aws-cdk-rfcs#state-diagram). I've converted this into draft mode to indicate that it is not yet ready for review.

**Comment by TheRealAmazonKendra** [comment]
> > New L2 constructs require an approved [RFC](https://github.com/aws/aws-cdk-rfcs) before a pull request will be reviewed. See our [state diagram for the RFC Process](https://github.com/aws/aws-cdk-rfcs#state-diagram). I've converted this into draft mode to indicate that it is not yet ready for review.
> 
> Shouldn't there already be approved RFCs if the previous two PRs made it to review?

I'm not sure when those were created but we've recently added this requirement.

**Comment by kaizencc** [comment]
@seanyles-amzn, I chatted with @TheRealAmazonKendra offline and we can actually use the prior PRs as a design doc for this L2, thereby satisfying the requirement for an RFC before implementation. Before I review, however, do you still have the bandwidth to iterate on this PR?

The first thing to do is to fix the failing build. Once that is done, please ping me and I will help review this PR!

**Comment by seanyles-amzn** [comment]
> @seanyles-amzn, I chatted with @TheRealAmazonKendra offline and we can actually use the prior PRs as a design doc for this L2, thereby satisfying the requirement for an RFC before implementation. Before I review, however, do you still have the bandwidth to iterate on this PR?
> 
> The first thing to do is to fix the failing build. Once that is done, please ping me and I will help review this PR!

Yes I still have the bandwidth to work this PR. I will update the build as soon as I can

**Comment by TheRealAmazonKendra** [comment]
Also, we recognize that some of the error messages emitted can be hard to understand. Please let us know if you need any clarification to get the build passing.

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by seanyles-amzn** [comment]
> Also, we recognize that some of the error messages emitted can be hard to understand. Please let us know if you need any clarification to get the build passing.

The linter check is simple enough to understand but the logs from CodeBuild are much too large for me to identify the issue. I could use some help on that front

**Comment by TheRealAmazonKendra** [comment]
Looks like the issue is that the props in a few of the interfaces aren't `readonly`.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Looks like something with this merge went quite wrong. In the future, please don't use force push and rebase with the command in console `@Mergifyio update`.

**Comment by seanyles-amzn** [comment]
I am git-illiterate aparently

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by seanyles-amzn** [comment]
> [awslint:no-unused-type:@aws-cdk/aws-cloudwatch.InsightRuleMetrics] type or enum is not used by exported classes (declared at lib/insight-rule.ts:11)

Not really sure what to do about this. This enum is used in the `metric` function of the InsightRule so that a user doesn't need to know the specific string metricName

**Comment by seanyles-amzn** [comment]
https://github.com/Mergifyio update

**Comment by seanyles-amzn** [comment]
> > [awslint:no-unused-type:@aws-cdk/aws-cloudwatch.InsightRuleMetrics] type or enum is not used by exported classes (declared at lib/insight-rule.ts:11)
> 
> Not really sure what to do about this. This enum is used in the `metric` function of the InsightRule so that a user doesn't need to know the specific string metricName

@TheRealAmazonKendra

**Comment by seanyles-amzn** [comment]
https://github.com/Mergifyio update

**Comment by seanyles-amzn** [comment]
> @aws-cdk/aws-cloudwatch: error: [awslint:resource-attribute:@aws-cdk/aws-cloudwatch.InsightRule.insightRuleRuleName] resources must represent all cloudformation attributes as attribute properties. "@attribute ATTR[,ATTR]" can be used to tag non-standard attribute names. missing property: insightRuleRuleName

I don't understand this @attribute ATTR[,ATTR] to change the propName

@TheRealAmazonKendra

File: packages/@aws-cdk/aws-cloudwatch/lib/insight-rule.ts [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+    }
+    return new Import(scope, id);
+  }
+
+  /**
+   * Name of this Insight Rule
+   *
+   * @attribute InsightRuleRuleName
+   */
+  public readonly insightRuleName: string;
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-cloudwatch/lib/insight-rule.ts:299]:
```suggestion
  public readonly insightRuleRuleName: string;
```

**Review by kaizencc — COMMENTED** [review]
@seanyles-amzn have you tried this? See [design guidelines](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#attributes) doc for more on attributes in CDK.

**Comment by seanyles-amzn** [comment]
```
@aws-cdk/aws-cloudwatch:   CHANGED    integ.alarm-and-dashboard 1.544s
@aws-cdk/aws-cloudwatch:       Resources
@aws-cdk/aws-cloudwatch: [~] AWS::CloudWatch::Dashboard DashCCD7F836 
@aws-cdk/aws-cloudwatch:  â””â”€ [~] DashboardBody
@aws-cdk/aws-cloudwatch:      â””â”€ [~] .Fn::Join:
@aws-cdk/aws-cloudwatch:          â””â”€ @@ -83,6 +83,10 @@
@aws-cdk/aws-cloudwatch:             [ ]         "QueueName"
@aws-cdk/aws-cloudwatch:             [ ]       ]
@aws-cdk/aws-cloudwatch:             [ ]     },
@aws-cdk/aws-cloudwatch:             [-]     "\"]],\"singleValueFullPrecision\":true}},{\"type\":\"custom\",\"width\":6,\"height\":6,\"x\":0,\"y\":56,\"properties\":{\"endpoint\":\"arn:aws:lambda:us-west-2:123456789012:function:my-function\",\"title\":\"My custom alarm\",\"updateOn\":{\"refresh\":true,\"resize\":true,\"timeRange\":true}}}]}"
@aws-cdk/aws-cloudwatch:             [+]     "\"]],\"singleValueFullPrecision\":true}},{\"type\":\"custom\",\"width\":6,\"height\":6,\"x\":0,\"y\":56,\"properties\":{\"endpoint\":\"arn:aws:lambda:us-west-2:123456789012:function:my-function\",\"title\":\"My custom alarm\",\"updateOn\":{\"refresh\":true,\"resize\":true,\"timeRange\":true}}},{\"type\":\"metric\",\"width\":6,\"height\":6,\"x\":0,\"y\":62,\"properties\":{\"insightRule\":{\"maxContributorCount\":10,\"orderBy\":\"Sum\",\"ruleName\":\"my-insight-rule\"},\"legend\":{\"position\":\"bottom\"},\"period\":300,\"region\":\"",
@aws-cdk/aws-cloudwatch:             [+]     {
@aws-cdk/aws-cloudwatch:             [+]       "Ref": "AWS::Region"
@aws-cdk/aws-cloudwatch:             [+]     },
@aws-cdk/aws-cloudwatch:             [+]     "\",\"title\":\"My Insight Widget\",\"view\":\"timeSeries\"}}]}"
@aws-cdk/aws-cloudwatch:             [ ]   ]
@aws-cdk/aws-cloudwatch:             [ ] ]
@aws-cdk/aws-cloudwatch:   UNCHANGED  integ.composite-alarm 1.593s
@aws-cdk/aws-cloudwatch:   UNCHANGED  integ.alarm-with-label 1.616s
@aws-cdk/aws-cloudwatch: Snapshot Results: 
@aws-cdk/aws-cloudwatch: Tests:    1 failed, 5 total
@aws-cdk/aws-cloudwatch: Error: Some tests failed!
```
This appears to be a snapshot that is failing because I updated it? How do I run the build to accept my changes?  I get `MODULE_NOT_FOUND` when trying `yarn integ-runner --update-on-failed` even after running `yarn install`

**Pull request closed without merging**